main.ts

```diff
@@ -0,0 +1,110 @@
+import { Octokit } from "@octokit/rest";
+import { parse, LockFileObject } from "@yarnpkg/lockfile";
+import { repos } from "./repos";
+
+async function main() {
+  if (process.argv.length < 3) {
+    printUsage();
+    process.exit(1);
+  }
+  const packagesToCheck = process.argv.splice(2);
+  const githubApi = setupGithubApi();
+  for (const packageNameParam of packagesToCheck) {
+    console.log(`Searching for package ${packageNameParam}`);
+    for (const repo of repos) {
+      console.log(`  In repo ${repo.repo}`);
+      const [org, repository] = repo.repo.split("/");
+      const lockFile = await getLockfile(
+        githubApi,
+        org,
+        repository,
+        repo.lockFile
+      );
+      const packageVersions = findVersions(lockFile, packageNameParam);
+      for (const packageName of Object.keys(packageVersions)) {
+        if (packageVersions[packageName].length > 0) {
+          console.log(
+            `    Package ${packageName}: ${packageVersions[packageName].join(
+              ", "
+            )}`
+          );
+        }
+      }
+    }
+  }
+}
+
+function printUsage() {
+  console.log(`Usage: `);
+  console.log(`  yarn list-versions <package-name>`);
+  console.log(`  yarn list-versions react`);
+  console.log(`  yarn list-versions react react-dom`);
+  console.log(`  yarn list-versions @kaizen/component-library`);
+  console.log(`  yarn list-versions "@kaizen/*"`);
+  console.log(`  yarn list-versions "@kaizen/*" "@cultureamp/*"`);
+}
+
+function setupGithubApi(): Octokit {
+  const token = process.env.GITHUB_TOKEN;
+  if (!token) {
+    throw `Please set process.env.GITHUB_TOKEN with a personal API token`;
+  }
+  return new Octokit({
+    auth: token,
+  });
+}
+
+async function getLockfile(
+  githubApi: Octokit,
+  org: string,
+  repo: string,
+  lockFilePath: string
+): Promise<LockFileObject> {
+  const result = await githubApi.repos.getContent({
```

Comment:
tfw you learn this API exists
![](https://thumbs.gfycat.com/AnyQueasyFairyfly-max-1mb.gif)

I'm curious as to whether (later, not needed now) being able to optionally specify the ref (branch, sha, whatever) in the repos.ts would be useful. I can't think of any case for _us_ right now, though, just cases where people have a 'stable' vs 'trunk' branch or something.

Reply:
You can! 

There's an optional `ref` parameter: https://developer.github.com/v3/repos/contents/#get-repository-content

A use case I can think of is to test if an `npm-refresh` branch has different versions than the `master` branch.

It could also be used to chart the versions over time (find the latest commit from X weeks ago and see when updates happenned... get a sense of the update lag)